### PR TITLE
abi generation

### PR DESCRIPTION
**Pull request by tbfleming**

### CODE REVIEW

File: libraries/eosiolib/contracts/eosio/dispatcher.hpp [hunk 1]
```diff
@@ -122,22 +166,36 @@ namespace eosio
...
-
 #define EOSIO_ACTION_WRAPPER_DECL(r, data, action)                                      \
    BOOST_PP_CAT(EOSIO_ACTION_WRAPPER_DECL_, BOOST_PP_COMPL(EOSIO_MATCH_NOTIFY(action))) \
    (r, data, action)
 
+#define EOSIO_REFLECT_ACTION_1(r, data, action)                                          \
+   f(BOOST_PP_CAT(                                                                       \
+       BOOST_PP_STRINGIZE(EOSIO_EXTRACT_ACTION_NAME(action)), _h),                       \
+       eosio::action_type_wrapper<&contract_class::EOSIO_EXTRACT_ACTION_NAME(action)>{}, \
+       EOSIO_GET_RICARDIAN(action) EOSIO_ACTION_ARG_NAME_STRINGS(action));
```
Comment (swatanabe) [inline libraries/eosiolib/contracts/eosio/dispatcher.hpp:182]: I tried giving an action a `ricadrian_contract` and it either vanished silently or was interpreted as an argument name, depending on the order.  In either case, the abigen program compiled and ran with no errors.

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+               return it->first;
+      }
+
+      template <typename Raw>
+      std::string get_type(bool force_alias = false)
+      {
+         using T = std::decay_t<Raw>;
+         if constexpr (is_std_vector<T>())
+         {
+            using inner = std::decay_t<typename is_std_vector<T>::value_type>;
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp:42]: That appears to have changed with C++11 and again with 17. `CopyAssignable` and `CopyConstructable` requirements on `T` were dropped and moved to individual vector methods.

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+                           ". Try EOSIO_REFLECT")
+                              .c_str());
+            }
+         }
+      }  // get_type
+
+      template <typename T, typename... Ts>
+      std::string generate_variant_name(std::variant<T, Ts...>* p)
+      {
+         return std::string{"variant<"} + (get_type<T>() + ... + ("," + get_type<Ts>())) + ">";
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp:122]: Yes. Needs fixing.

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (swatanabe) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp:42]: It looks like you're right.  `remove_cv` would be more appropriate, though.  Array to pointer decay is definitely not the right thing to do.

[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp:42]: `remove_cvref` (new in 20)

File: libraries/eosiolib/contracts/eosio/dispatcher.hpp [hunk 4]
```diff
@@ -81,24 +82,67 @@ namespace eosio
...
-#define EOSIO_EXTRACT_NOTIFY_CODE(x) BOOST_PP_CAT(EOSIO_EXTRACT_NOTIFY_CODE_, x)
-#define EOSIO_EXTRACT_NOTIFY_CODE_notify(code, action) code
-#define EOSIO_EXTRACT_NOTIFY_ACTION(x) BOOST_PP_CAT(EOSIO_EXTRACT_NOTIFY_ACTION_, x)
-#define EOSIO_EXTRACT_NOTIFY_ACTION_notify(code, action) action
-
-#define EOSIO_DISPATCH_ACTION_INTERNAL_1(r, type, member)                                        \
-   case eosio::hash_name(BOOST_PP_STRINGIZE(member)):                                            \
-      executed = eosio::execute_action(eosio::name(receiver), eosio::name(code), &type::member); \
+#define EOSIO_MATCH_ACTION(x) EOSIO_MATCH(EOSIO_MATCH_ACTION, x)
+#define EOSIO_MATCH_ACTIONaction EOSIO_MATCH_YES
```
[the remark this answers is not in the document]
Comment (swatanabe) [reply libraries/eosiolib/contracts/eosio/dispatcher.hpp:102]: Ah.  I didn't look at EOSIO_MATCH_CHECK closely enough.

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: changed

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: changed

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+         }
+         d.types.push_back(name);
+         add_variant_types(d, (type_list<Ts...>*)nullptr, names...);
+      }
+
+      template <typename T, typename... Ts>
+      void add_variant_types(variant_def& d, type_list<T, Ts...>*)
+      {
+         d.types.push_back(get_type<T>());
+         add_variant_types(d, (type_list<Ts...>*)nullptr);
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: changed

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+               auto it = type_to_name.find(type);
+               if (it != type_to_name.end())
+                  return it->second;
+               auto inner_name = get_type<inner>(true);
+               const auto& name = reserve_name("vector<" + inner_name + ">", type);
+               type_to_name[typeid(T)] = name;
+               def.types.push_back({name, inner_name + "[]"});
+               return name;
+            }
+            return get_type<inner>(is_std_vector<inner>() || is_std_optional<inner>()) + "[]";
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: changed

File: libraries/eosiolib/contracts/eosio/dispatcher.hpp (hunk 1 above)
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/dispatcher.hpp:182]: Odd. I moved it around in token and it generated correctly:

```
   EOSIO_ACTIONS(contract,
                 "eosio.token"_n,
                 action(create, ricardian_contract(create_ricardian), issuer, maximum_supply),
                 action(issue, to, ricardian_contract(issue_ricardian), quantity, memo),
                 action(retire, quantity, memo, ricardian_contract(retire_ricardian)),
                 action(transfer, from, to, ricardian_contract(transfer_ricardian), quantity, memo),
                 action(open, owner, symbol, ram_payer, ricardian_contract(open_ricardian)),
                 action(close, owner, symbol, ricardian_contract(close_ricardian)))
```

Comment (swatanabe) [reply libraries/eosiolib/contracts/eosio/dispatcher.hpp:182]: Look at the spelling of my test more closely.

Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/dispatcher.hpp:182]: ah. Probably need to add a check for unexpected parens.

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (swatanabe) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: What about the first type?  Also, we only need a different name, we don't actually need to define a new type, right?

File: libraries/eosiolib/contracts/eosio/dispatcher.hpp (hunk 1 above)
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/dispatcher.hpp:182]: Added checks for unused action arg names and for unexpected `()` within arg names.

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: changed

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+            if (auto [it, inserted] = name_to_type.insert({base + std::to_string(i), type});
+                inserted)
+               return it->first;
+      }
+
+      template <typename Raw>
+      std::string get_type(bool force_alias = false)
+      {
+         using T = std::decay_t<Raw>;
+         if constexpr (is_std_vector<T>())
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: Added several more. `map` won't work until `pair` is added.

File: libraries/eosiolib/contracts/eosio/abi_generator.hpp [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+   gen.def.ricardian_clauses.push_back({         \
+       EOSIO_ABIGEN_EXTRACT_CLAUSE_ID(clause),   \
+       EOSIO_ABIGEN_EXTRACT_CLAUSE_BODY(clause), \
+   });
+
+#define EOSIO_ABIGEN_ITEM(r, data, item)                                                    \
+   BOOST_PP_IIF(EOSIO_ABIGEN_MATCH_ACTIONS(item), EOSIO_ABIGEN_ACTION, EOSIO_EMPTY)(item);  \
+   BOOST_PP_IIF(EOSIO_ABIGEN_MATCH_TABLE(item), EOSIO_ABIGEN_TABLE, EOSIO_EMPTY)(item);     \
+   BOOST_PP_IIF(EOSIO_ABIGEN_MATCH_VARIANT(item), EOSIO_ABIGEN_VARIANT, EOSIO_EMPTY)(item); \
+   BOOST_PP_IIF(EOSIO_ABIGEN_MATCH_CLAUSE(item), EOSIO_ABIGEN_CLAUSE, EOSIO_EMPTY)(item);
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: Yep. That simplified it a lot.

[the remark this answers is not in the document]
Comment (swatanabe) [reply libraries/eosiolib/contracts/eosio/abi_generator.hpp]: Could use `static_assert` instead of `check` for unknown items.

**Review by swatanabe — APPROVED** [review]

**Pull request merged**